eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,231 @@
+                                amount: MilliSatoshi,
+                                expiry: CltvExpiry,
+                                preimage: ByteVector32,
+                                customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override val totalAmount = amount
+  override val features = Features.empty
+  override val extraEdges = Nil
+
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    Recipient.verifyChannelHopsOnly(route).map(hops => {
+      val finalPayload = NodePayload(nodeId, FinalPayload.Standard.createKeySendPayload(route.amount, amount, expiry, preimage, customTlvs))
+      OutgoingPaymentPacket.buildPayloads(amount, expiry, finalPayload, hops)
+    })
+  }
+}
+
+/** A payment recipient that hides its real identity using route blinding. */
+case class BlindedRecipient(nodeId: PublicKey,
+                            features: Features[InvoiceFeature],
+                            totalAmount: MilliSatoshi,
+                            expiry: CltvExpiry,
+                            extraEdges: Seq[Invoice.BlindedEdge],
+                            customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    extractRoute(route).map { case (channelHops, blindedHop) =>
+      val introductionPayload = NodePayload(blindedHop.route.introductionNodeId, OutgoingBlindedPerHopPayload.createIntroductionPayload(blindedHop.route.introductionNode.encryptedPayload, blindedHop.route.introductionNode.blindingEphemeralKey))
+      val intermediateBlindedPayloads = blindedHop.route.subsequentNodes.dropRight(1).map(n => NodePayload(n.blindedPublicKey, OutgoingBlindedPerHopPayload.createIntermediatePayload(n.encryptedPayload)))
+      val finalBlindedPayload = NodePayload(blindedHop.route.blindedNodes.last.blindedPublicKey, OutgoingBlindedPerHopPayload.createFinalPayload(route.amount, totalAmount, expiry, blindedHop.route.blindedNodes.last.encryptedPayload, customTlvs))
+      val introductionAmount = route.amount + blindedHop.paymentInfo.fee(route.amount)
+      val introductionExpiry = expiry + blindedHop.paymentInfo.cltvExpiryDelta
+      val blindedPayloads = PaymentPayloads(introductionAmount, introductionExpiry, introductionPayload +: intermediateBlindedPayloads :+ finalBlindedPayload)
+      if (channelHops.isEmpty) {
+        // We are the introduction node of the blinded route.
+        blindedPayloads
+      } else {
+        channelHops.tail.reverse.foldLeft(blindedPayloads) {
+          case (current, channelHop) =>
+            val payload = NodePayload(channelHop.nodeId, IntermediatePayload.ChannelRelay.Standard(channelHop.shortChannelId, current.amount, current.expiry))
+            PaymentPayloads(current.amount + channelHop.fee(current.amount), current.expiry + channelHop.cltvExpiryDelta, payload +: current.payloads)
```

Comment:
Not exactly, because we start with a sequence of payloads (for all blinded hops) instead of just a final payload. But you're right that we can easily change `OutgoingPaymentPacket.buildPayloads` to take a `finalPayloads: Seq[NodePayload]` to reuse it here, I'll do that.

Reply:
Fixed in the latest version.